sys/include/net/gnrc/netif2.h

```diff
@@ -0,0 +1,322 @@
+ *
+ * Network interfaces in the context of GNRC are threads for protocols that are
+ * below the network layer.
+ *
+ * @{
+ *
+ * @file
+ * @brief   Definition for GNRC's network interfaces
+ *
+ * @author  Martine Lenders <mlenders@inf.fu-berlin.de>
+ */
+#ifndef GNRC_NETIF2_H
+#define GNRC_NETIF2_H
+
+#include <stddef.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+#include "kernel_types.h"
+#include "msg.h"
+#include "net/gnrc/netapi.h"
+#include "net/gnrc/pkt.h"
+#include "net/gnrc/netif2/conf.h"
+#ifdef MODULE_GNRC_SIXLOWPAN
+#include "net/gnrc/netif2/6lo.h"
+#endif
+#include "net/gnrc/netif2/flags.h"
+#ifdef MODULE_GNRC_IPV6
+#include "net/gnrc/netif2/ipv6.h"
+#endif
+#ifdef MODULE_GNRC_MAC
+#include "net/gnrc/netif2/mac.h"
+#endif
+#include "net/netdev.h"
+#include "rmutex.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Operations to an interface
+ */
+typedef struct gnrc_netif2_ops gnrc_netif2_ops_t;
+
+/**
+ * @brief   Representation of a network interface
+ */
+typedef struct {
+    const gnrc_netif2_ops_t *ops;           /**< Operations of the network interface */
+    netdev_t *dev;                          /**< Network device of the network interface */
+    rmutex_t mutex;                         /**< Mutex of the interface */
+#if defined(MODULE_GNRC_IPV6) || DOXYGEN
+    gnrc_netif2_ipv6_t ipv6;                /**< IPv6 component */
+#endif
+#if defined(MODULE_GNRC_MAC) || DOXYGEN
+    gnrc_netif2_mac_t mac;                  /**< @ref gnrc_mac component */
+#endif  /* MODULE_GNRC_MAC */
+    /**
+     * @brief   Flags for the interface
+     *
+     * @see net_gnrc_netif2_flags
+     */
+    uint32_t flags;
+#if (GNRC_NETIF2_L2ADDR_MAXLEN > 0)
+    /**
+     * @brief   The link-layer address currently used as the source address
+     *          on this interface.
+     *
+     * @note    Only available if @ref GNRC_NETIF2_L2ADDR_MAXLEN > 0
+     */
+    uint8_t l2addr[GNRC_NETIF2_L2ADDR_MAXLEN];
+
+    /**
+     * @brief   Length in bytes of gnrc_netif2_t::l2addr
+     *
+     * @note    Only available if @ref GNRC_NETIF2_L2ADDR_MAXLEN > 0
+     */
+    uint8_t l2addr_len;
+#endif
+#if defined(MODULE_GNRC_SIXLOWPAN) || DOXYGEN
+    gnrc_netif2_6lo_t sixlo;                /**< 6Lo component */
+#endif
+    uint8_t cur_hl;                         /**< Current hop-limit for out-going packets */
+    uint8_t device_type;                    /**< Device type */
+    kernel_pid_t pid;                       /**< PID of the network interface's thread */
+} gnrc_netif2_t;
+
+/**
+ * @see gnrc_netif2_ops_t
+ */
+struct gnrc_netif2_ops {
+    /**
+     * @brief   Initializes network interface beyond the default settings
+     *
+     * @pre `netif != NULL`
+     *
+     * @param[in] netif The network interface.
+     *
+     * This is called after the default settings were set, right before the
+     * interface's thread starts receiving messages. It is not necessary lock
```

Comment:
'It is not necessary lock the interface's mutex gnrc_netif_t::mutex, since the thread will already lock it.'
I think there's a typo here, that's why I proposed:
'It doesn't necessary lock...'
Maybe some rephrasing is required.